checker/cost.go

```diff
@@ -0,0 +1,536 @@
+package checker
+
+import (
+	"fmt"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"math"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/google/cel-go/checker/decls"
+	"github.com/google/cel-go/common/containers"
+	"github.com/google/cel-go/common/types/ref"
+)
+
+// CostEstimator estimates the sizes of variable length input data and the costs of functions.
+type CostEstimator interface {
+	// EstimateSize returns the worst case size possible for the given Expr, or nil if
+	// the estimator has no estimate to provide. EstimateSize is only called for Exprs where
+	// CEL does not know the size; EstimateSize is not called for literals defined inline in CEL.
+	EstimateSize(element Expr) *int64
+	// EstimateCallCost returns the estimated cost of an invocation, or nil if
+	// the estimator has no estimate to provide. EstimateCallCost is only called for functions
+	// that CEL does not already know the cost; EstimateCallCost is not called for standard operations
+	// or functions.
+	EstimateCallCost(name string, target *Expr, args []Expr) *CostEstimate
+}
+
+// Expr represents an AST node for the purpose of cost estimations.
+type Expr interface {
+	// Type returns the deduced type of the Expr.
+	Type() *exprpb.Type
+	// LiteralSize returns the size of the Expr if it is a literal defined inline in CEL, or
+	// nil if the Exprs size is not statically known.
+	LiteralSize() *int64
+}
+
+type expr struct {
+	t    *exprpb.Type
+	expr *exprpb.Expr
+}
+
+func (e expr) Type() *exprpb.Type {
+	return e.t
+}
+
+func (e expr) LiteralSize() *int64 {
+	var v int64
+	switch ek := e.expr.ExprKind.(type) {
+	case *exprpb.Expr_ConstExpr:
+		switch ck := ek.ConstExpr.ConstantKind.(type) {
+		case *exprpb.Constant_StringValue:
+			v = int64(len(ck.StringValue))
+		case *exprpb.Constant_BytesValue:
+			v = int64(len(ck.BytesValue))
+		default:
+			return nil
+		}
+	case *exprpb.Expr_ListExpr:
+		v = int64(len(ek.ListExpr.Elements))
```

Comment:
It's worth a comment about how list size is used here. 

The list _could_ be the size of whatever elements it contains. For example a set membership test of `string in [a, b, c]` it might be relevant to know if either `a`, `b`, or `c` were quite large. It's very probable that you've accounted for this elsewhere, but in the event you haven't, then it's worth spending more time to hammer out.

Reply:
I've clarified that size is the same as what CEL `size()` computes (lengths of strings/bytes, # of list items, # of map entries).